### ISSUE TITLE
Inconsistent reads when reading from Buffer table engine

### ISSUE THREAD

**Issue by luca-nardelli** [ClickHouse/ClickHouse#44766]
**Describe the unexpected behaviour**

**Note: We are running on clickhouse cloud on a v22.11 cluster**

In order to work around a bug on async inserts, which should have been fixed by https://github.com/ClickHouse/ClickHouse/pull/43233 , we started experimenting with Buffer tables. However we noticed that when trying to read from the Buffer table, we get inconsistent reads with rows seemingly missing from some results.

Main table structure

```sql
CREATE TABLE evm_transactions
(
    `timestamp`    DateTime64(3),   --- Milliseconds
    `chain_id`     LowCardinality(String),
    `block_number` UInt256,
    `tx_index`     UInt32,
    `version`      DateTime DEFAULT NOW(),
) ENGINE = ReplacingMergeTree(version)
ORDER BY (timestamp, block_number, tx_index, chain_id)
PRIMARY KEY (timestamp)
PARTITION BY toYYYYMM(timestamp)
SETTINGS index_granularity=1024
;
```

Buffer table

```sql
CREATE TABLE buf_test_evm_transactions_nopart AS evm_transactions
ENGINE = Buffer(
	currentDatabase(),
	test_evm_transactions_nopart,
	4, -- num_layers
	120, -- min_time
	600, -- max_time
	500, -- min_rows
	10000, -- max_rows
	10000000, -- min_bytes
	100000000 -- max_bytes
)
```

Test query

```sql
SELECT NOW(),timestamp, chain_id, block_number, tx_index
FROM buf_test_evm_transactions_nopart
WHERE chain_id = '43114'
AND timestamp > NOW() - 60
ORDER BY timestamp DESC, block_number DESC, tx_index DESC
LIMIT 10;
```

If we repeat this query from our db client (in this specific example this happens with DBeaver) in a very short amount of time we get changing data, as if we were looking at different buffers that old parts of the data. I am attaching an example video below

https://user-images.githubusercontent.com/6215162/210100226-4720b4cd-05f0-4998-8de5-8bc5805f6c6b.mp4

Is this intended behavior? I understand that buffers are spread out over different nodes of the cluster, however I couldn't find a reference to this behavior in the documentation (https://clickhouse.com/docs/en/engines/table-engines/special/buffer/). Our goal was to leverage buffer tables with a high enough buffers to group data, while still being able to query **all of it** within clickhouse (our event streams right now are ~ 50 events/sec, so we would like to use relatively long time buffers).

**Comment by den-crane** [comment]
It is very hard to understand what you mean watching this video. I wasted 5 minutes. Understood nothing.
There is no proof that you query the same one server. If you use something like K8s / load balancer it could be that you query different data.

Avoid to query engine buffer, it has tons of unexpected behaviors especially with  ReplacingMergeTree behind.
Query real table `evm_transactions`.

>I understand that buffers are spread out over different nodes of the cluster

No. Nothing like this. Buffer is just a buffer at the current node.

**Comment by luca-nardelli** [comment]
Thanks for your answer. First of all, let me say that your writing style comes out as a little arrogant. I understand my question could come from inexperience and that could frustrate you, but it seems very uncalled for.

----

Now, to answer your points.

>There is no proof that you query the same one server. If you use something like K8s / load balancer it could be that you query different data.

My first line in the description above states that we are running this on [Clickhouse Cloud](https://clickhouse.com/cloud). I don't know how the setup is there, but it's likely there is some kind of load balancer in between.
To add more information to the video above, I was connecting using JDBC with DBeaver as a client with the standard connection string given to us by Clickhouse Cloud, and I was repeating the same query multiple times over a short time span, observing that the results changed. Unfortunately that was what I thought was the best way to show the issue, but it seems it wasn't enough. I don't think that was related to the "NOW()" usage since the same data would repeat after a couple tries.
Moreover, the block_number should be a contiguous range of numbers, whereas in the video we can clearly see gaps.

>Avoid to query engine buffer, it has tons of unexpected behaviors especially with ReplacingMergeTree behind and DEFAULT NOW().
Query real table evm_transactions.

This is all very unclear.  "tons of unexpected behavior" does not allow me to make an informed decision and decide whether to query the buffer table or the real table. Do you have any resources I could look at to understand these undefined behaviors more? The only issues I could find here (https://clickhouse.com/docs/en/engines/table-engines/special/buffer) were related to loss of deduplication when writing to replicated tables.

The whole point of using the Buffer table was to buffer our inserts (that are coming in at ~50 rows/sec) and insert them in fairly large batches (to reduce parts, since having many parts leads to latency in reads) for example every couple of minutes, **while still being able to query realtime data as it is coming in.**

>No. Nothing like this. Buffer is just a buffer at the current node.

So, how does Buffer engine work when running a multi-node setup (which I assume is the case for Clickhouse Cloud)?  According to your message, it seems that only the data on the current node is returned. If this is the case, documentation should be updated because it looks to me that it doesn't say anywhere that in a multi-node setup only a subset of the data is returned.

Relevant excerpt from the docs:
>When reading from a Buffer table, data is processed both from the buffer and from the destination table (if there is one). Note that the Buffer table does not support an index. In other words, data in the buffer is fully scanned, which might be slow for large buffers. (For data in a subordinate table, the index that it supports will be used.)

**Comment by den-crane** [comment]
>Thanks for your answer. First of all, let me say that your writing style comes out as a little arrogant. I understand my question could come from inexperience and that could frustrate you, but it seems very uncalled for.

I agree that that my style is unappropriate. I am very toxic person. I am trying to be more polite. I will try to turn off toxicity for 5 minutes.
Be aware that I am not a part of Clickhouse team, I have no relation to Clickhouse, inc. So treat me as a random guy who is trying to help you.

BTW, you showed DDL for `evm_transactions` but `buf_test_evm_transactions_nopart` uses another table **test_evm_transactions_nopart**


>My first line in the description above states that we are running this on [Clickhouse Cloud](https://clickhouse.com/cloud).

Sorry I missed that. I think you query different Clickhouse servers and they see different buffers and different replicas. Because of this you get different results. Buffer engine is not replicated. Replicated tables has a replication lag. So with and without buffer results of the queries will inconsistent. If you need consistent results then Buffer engine is definitely no go.

>This is all very unclear. "tons of unexpected behavior"

ReplacingMergeTree has a special property -- it deduplicates/replaces rows by key during the insertion and during merges. Buffer engine does not. 

Here is an example:
```sql
CREATE TABLE t ( key Int64, s String , `version` DateTime DEFAULT NOW())
ENGINE = ReplacingMergeTree(version) PRIMARY KEY key ORDER BY (key);

CREATE TABLE buf_t AS t
ENGINE = Buffer(	currentDatabase(), 	t,
	4, -- num_layers
	12000, -- min_time
	60000, -- max_time
	500, -- min_rows
	10000, -- max_rows
	10000000, -- min_bytes
	100000000 -- max_bytes
);

insert into buf_t values(1, '1', now());
insert into buf_t values(1, '2', now());

select * from buf_t final;

┌─key─┬─s─┬─────────────version─┐
│   1 │ 2 │ 2022-12-30 21:39:45 │
│   1 │ 1 │ 2022-12-30 21:39:38 │
└─────┴───┴─────────────────────┘

detach table buf_t; -- emulate buffer flush
attach table buf_t; -- emulate buffer flush

select * from buf_t final;
┌─key─┬─s─┬─────────────version─┐
│   1 │ 2 │ 2022-12-30 21:44:38 │
└─────┴───┴─────────────────────┘
```

as you can see result has changed after the flush to MergeTree table.

I think you should not query Engine=Buffer table. 
And you should reduce Engine=Buffer max_time to 1 or 2 seconds if you need more realtime results.

**Comment by luca-nardelli** [comment]
>I agree that that my style is unappropriate. I am very toxic person. I am trying to be more polite. I will try to turn off toxicity for 5 minutes.
Be aware that I am not a part of Clickhouse team, I have no relation to Clickhouse, inc. So treat me as a random guy who is trying to help you.

I appreciate your frankness :smile: thank you!

>BTW, you showed DDL for evm_transactions but buf_test_evm_transactions_nopart uses another table test_evm_transactions_nopart.

My bad, I was trying to extract bits and pieces from our tests and missed that.

>Buffer engine is not replicated.

I think this is the key part/issue then. Since I guess Clickhouse Cloud works with multiple nodes, creating a Buffer table with N layers effectively creates N*M buffers (where M is the number of nodes), and whenever we query we hit a random node leading to a different subset of the buffers being read.

>ReplacingMergeTree has a special property -- it deduplicates/replaces rows by key during the insertion and during merges. Buffer engine does not.

Yeah you're right, I was aware of that, I thought there were other issues I was ignoring so I asked. Thanks for the example though!

> I think you should not query Engine=Buffer table.

I agree with you, it's basically useless in a replicated scenario I think, unless you are ok with "sampling" recent data instead of showing all of it. The documentation should probably have a warning or something to tell that to devs.

>And you should reduce Engine=Buffer max_time to 1 or 2 seconds if you need more realtime results.

In another experiment we used a short lived buffer (3 seconds), and that led to CH constantly having ~10 small parts all within the same 1-2 minute window, which led to high latency in queries that had to retrieve recent data since CH always has to read from all parts even when reading in order (this is exacerbated also by the fact that parts are on S3 in CH Cloud). I suppose we'll probably use another system entirely for "live" data and use CH for older data, so that we can use bigger buffers to keep the parts under control

A ReplicatedBuffer/DistributedBuffer engine would be cool though :D

**Comment by den-crane** [comment]
>In another experiment we used a short lived buffer (3 seconds), and that led to CH constantly having ~10 small parts all within the same 1-2 minute window, which led to high latency in queries that had to retrieve recent data since CH always has to read from all parts even when reading in order (this is exacerbated also by the fact that parts are on S3 in CH Cloud).

Probably it something related to the Cloud / S3 cache.
In general on-premises local disks there is no any visible latency if the number of parts in a partition less than 100.

**Comment by luca-nardelli** [comment]
That's very good to know, thanks! Also, I agree, it's definitely related to S3.

By the way, as a reference for future readers, I just discovered that I can query the buffer table with `clusterAllReplicas` function (`FROM clusterAllReplicas(default,<dbname>.buf_evm_transactions)`  and I get consistent readings from the buffers then. By also adding `hostName()` to the selection list I am able to see which node has which data row. However this process incurs extra network roundtrips because we need to read from all the nodes in the cluster so I guess in our case it's not worth it.

I'll be closing this issue as it's something that works as expected and I found the proper way of querying those tables.
